### PR TITLE
Reduce direct locomotion step overhead

### PR DESCRIPTION
**Pull request by AntoineRichard**
# Description

Reduces common Direct Locomotion task overhead used by Ant and Humanoid without introducing the experimental Ant post-step path.

The change:

- stages effort targets once per environment step because articulation command buffers persist across decimation substeps;
- uses the cached all-joint target path instead of repeatedly uploading joint indices;
- keeps scalar zero assignments on-device;
- removes a duplicate articulation reset already performed by the scene reset;
- removes redundant clones after advanced indexing; and
- uses the common fused joint-state writer provided by PhysX, Newton, and Isaac Lab OV.

The isolated device scalar change improved the Ant host-return measurement from 3.433 ms to 3.231 ms. The remaining changes remove repeated launches, index transfers, copies, and reset writes; the larger fused post-step and device-mask work is deliberately excluded.

This is one focused slice of superseded Draft PR #6509 and is independent of benchmark PR #6474.

## Type of change

- Bug fix (non-breaking change which fixes an issue)

## Release backport

- [x]  Backport this pull request to the active release branch after it merges into `develop`

## Screenshots

Not applicable.

## Validation

- `uv run python tools/changelog/cli.py check develop`
- `uv run --frozen isaaclab -f`
- `uv run --frozen --extra test python -m pytest source/isaaclab_tasks/test/core/test_environments_newton.py -k 'Direct and (Ant or Humanoid)' -q` (`2 passed`)
- Verified the fused writer interface on PhysX, Newton, and Isaac Lab OV.
- The source change was exercised by the original focused CPU/CUDA validation before the PR split; new test files are intentionally omitted from this slice.

## Checklist

Docker and GPU tests run on demand. Push the commits you want tested, then
comment `run-ci` on the pull request.

- [x] I have read and understood the [contribution guidelines](https://isaac-sim.github.io/IsaacLab/main/source/refs/contributing.html)
- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `./isaaclab.sh --format`
- [x] I have made corresponding changes to the documentation (not required for this internal performance fix)
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added a changelog fragment under `source/<pkg>/changelog.d/` for every touched package (do **not** edit `CHANGELOG.rst` or bump `extension.toml` — CI handles that)
- [x] I have added my name to the `CONTRIBUTORS.md` or my name already exists there

### CODE REVIEW

File: source/isaaclab_tasks/isaaclab_tasks/core/locomotion/locomotion_direct_env.py [hunk 1]
```diff
@@ -232,24 +233,27 @@ def _reset_idx(self, env_ids: torch.Tensor | None):
 
         # Log survival success rate (survived = timed out without falling)
         survived = self.reset_time_outs[env_ids].float()
-        self.extras.setdefault("log", {})["Metrics/success_rate"] = survived.mean().item()
+        self.extras.setdefault("log", {})["Metrics/success_rate"] = survived.mean()
```
[the remark this answers is not in the document]
Comment (AntoineRichard) [reply source/isaaclab_tasks/isaaclab_tasks/core/locomotion/locomotion_direct_env.py]: Good catch. Fixed in 08ac26539c2 by restoring .item(), which preserves the historical Python-float contract for extras["log"] and SB3-style episode info.

File: source/isaaclab_tasks/isaaclab_tasks/core/locomotion/locomotion_direct_env.py [hunk 2]
```diff
@@ -132,10 +131,12 @@ def _setup_scene(self):
 
     def _pre_physics_step(self, actions: torch.Tensor) -> None:
         self.actions = actions.clone()
+        forces = self.action_scale * self.joint_gears * torch.clamp(self.actions, -1.0, 1.0)
```
[the remark this answers is not in the document]
Comment (AntoineRichard) [reply source/isaaclab_tasks/isaaclab_tasks/core/locomotion/locomotion_direct_env.py:134]: I verified command lifetime across Newton, PhysX, and OVPhysX and kept the staging change. set_joint_effort_target_index writes persistent articulation command buffers; scene.write_data_to_sim reapplies them on every decimation substep, and none of the three backends consumes or clears the buffer after stepping. The focused action-staging coverage passes.

File: source/isaaclab_tasks/isaaclab_tasks/core/locomotion/locomotion_direct_env.py [hunk 3]
```diff
@@ -232,24 +233,27 @@ def _reset_idx(self, env_ids: torch.Tensor | None):
...
 
-        joint_pos = self.robot.data.default_joint_pos.torch[env_ids].clone()
-        joint_vel = self.robot.data.default_joint_vel.torch[env_ids].clone()
-        default_root_pose = self.robot.data.default_root_pose.torch[env_ids].clone()
-        default_root_vel = self.robot.data.default_root_vel.torch[env_ids].clone()
+        joint_pos = self.robot.data.default_joint_pos.torch[env_ids]
+        joint_vel = self.robot.data.default_joint_vel.torch[env_ids]
+        default_root_pose = self.robot.data.default_root_pose.torch[env_ids]
+        default_root_vel = self.robot.data.default_root_vel.torch[env_ids]
         default_root_pose[:, :3] += self.scene.env_origins[env_ids]
```
[the remark this answers is not in the document]
Comment (AntoineRichard) [reply source/isaaclab_tasks/isaaclab_tasks/core/locomotion/locomotion_direct_env.py:244]: I checked this against the actual reset control flow and kept the optimization. _reset_idx normalizes full resets to wp.to_torch(self.robot._ALL_INDICES), while partial resets already arrive as integer tensors. PyTorch tensor advanced indexing with env_ids returns separate storage, so the in-place origin addition cannot mutate default_root_pose. The focused reset immutability coverage passes.

File: source/isaaclab_tasks/isaaclab_tasks/core/locomotion/locomotion_direct_env.py [hunk 4]
```diff
@@ -234,22 +235,25 @@ def _reset_idx(self, env_ids: torch.Tensor | None):
...
 
         self.robot.write_root_pose_to_sim_index(root_pose=default_root_pose, env_ids=env_ids)
         self.robot.write_root_velocity_to_sim_index(root_velocity=default_root_vel, env_ids=env_ids)
-        self.robot.write_joint_position_to_sim_index(position=joint_pos, env_ids=env_ids)
-        self.robot.write_joint_velocity_to_sim_index(velocity=joint_vel, env_ids=env_ids)
+        if hasattr(self.robot, "write_joint_state_to_sim_index"):
```
Comment (ooctipus) [inline source/isaaclab_tasks/isaaclab_tasks/core/locomotion/locomotion_direct_env.py]: hasattr(self.robot, "write_joint_state_to_sim_index") <----------- is this for ovphysx? because physx and newton both provide write_joint_state_to_sim_index, why ovphysx doesn't provide?

Comment (AntoineRichard) [reply source/isaaclab_tasks/isaaclab_tasks/core/locomotion/locomotion_direct_env.py]: Yes, this was the OVPhysX compatibility path. PR #6513 has now merged and all three backends provide write_joint_state_to_sim_index, so I removed the hasattr/split-writer fallback in ff80157b475.

**Review by ooctipus — APPROVED** [review]

**Review by hujc7 — APPROVED** [review]

**Comment by AntoineRichard** [comment]
run-ci
